### DISCUSSION TITLE
Warning when use simulate_waveforms()

### DISCUSSION THREAD

**Discussion by zql2021** [JuliaPhysics/SolidStateDetectors.jl#376] [Q&A] [answered]
Hello,Thank you for your previous patient answers, which helped me a lot.
Recently,when using  `simulate_waveforms`,  I encountered numerous warning, primarily  three types:
```
Warning: 3 charges of the charge cloud at Float32[-0.003572, 0.000623, 0.009535] are outside. Moving them inside...
The new charge center is at Float32[-0.003572, 0.000623, 0.009532].

Warning: Internal error for charge starting at Float32[-0.0084860185, 0.0073255054, 0.001050136]

Warning: Handling of charge at floating boundary did not work as intended. Start Position (Cart): Float32[0.011792993, 0.009161911, 0.009117997]
```
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/58914476/41091461-137e-4ad3-abf5-23925a37de9e)
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/58914476/1382ac15-528a-45d8-8889-0f9da59b66cc)
Even refer to the src code, I'm still unclear about the cause of these warnings. Do they have a bad influence on the final signals (waveforms), How can I avoid these warnings? I want to seek your advice on this matter.
Below is my main code, hoping it  help you understand m y issue better.
```
simulate!(sim, refinement_limits = [0.3, 0.1, 0.05], 
max_tick_distance = 0.2u"mm",convergence_limit=1e-6,
max_distance_ratio = 4,verbose=true)

mc_events_clustered = @time SolidStateDetectors.cluster_detector_hits(mc_events, 0.1u"mm")

det_material = sim.detector.semiconductor.material
mc_events_fnoise = add_fano_noise(mc_events_det1, det_material.E_ionisation, det_material.f_fano)

filtered_events = mc_events_fnoise[findall(pts -> all(p -> CartesianPoint(T.(ustrip.(uconvert.(u"m", p)))) in sim.detector, pts), mc_events_fnoise.pos)];

contact_charge_signals = simulate_waveforms(      
        filtered_events[1:1000],
        sim,
        max_nsteps = 4000, 
        Δt = 1u"ns", 
        number_of_carriers = 100, 
        diffusion = true, 
        self_repulsion = true, 
        verbose=true);
```
`plot(filtered_events)` get 
![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/58914476/7ff8affc-4467-4293-a89e-3a8a376455d6)

and my grid setting is 
```
grid:
  coordinates: cartesian
  axes:
    x:
      from: -16
      to: 16
      boundaries: inf
    y:
      from: -16
      to: 16
      boundaries: inf
    z:
      from: -1
      to: 11
      boundaries: inf
```
the same waring appear when using a cylinder coordinate system.

**Comment by fhagemann** [comment, marked as answer]
I must admit that the warning messages might not be very responsive, so let me try to explain the origin of these three warning types and whether they have an influence on the quality of the final waveforms:

```
Warning: 3 charges of the charge cloud at Float32[-0.003572, 0.000623, 0.009535] are outside. Moving them inside...
The new charge center is at Float32[-0.003572, 0.000623, 0.009532].
```
In order to simulate diffusion and self-repulsion of the charge carriers, the charge clouds need to consist of multiple charge carriers.
You are using `number_of_carriers = 100`, meaning that each energy deposit will be modeled as 100 point charges, evenly distributed on surfaces of spheres, see [here](https://juliaphysics.github.io/SolidStateDetectors.jl/stable/man/charge_drift/#Equally-Distributed-on-a-Regular-Sphere). If the energy deposit is very close to a surface of the detector, some point charges in the initial charge cloud configuration would be outside of the semiconductor. If this is the case, the charges falling outside of the detector volume are moved towards the center of the charge cloud until they are back in. This can effectively shift the center of the charge cloud a bit more inside. In this case, it was moved 3µm down (that should not affect the waveforms too much).

```
Warning: Internal error for charge starting at Float32[-0.0084860185, 0.0073255054, 0.001050136]
```
Charges are drifted, ideally until they reach a contact. Whenever a charge would leave the semiconductor, we determine the intersect of the last drift step vector with the surface of the semiconductor. Then we check, whether this intersection is on a contact, an uncontacted surface of the semiconductor or "something else". Usually, we would not expect "something else", because the surface should either be contacted or uncontacted. However, due to rounding errors, the charges might end up slightly in- or outside. In that case, the charge is still "in the bulk" or "outside". In that case, we still end the drift but throw the warning.
However, as the distance to the surface might again be of the order of µm, it should not affect the waveforms significantly and this warning can also be ignored.

```
Warning: Handling of charge at floating boundary did not work as intended. Start Position (Cart): Float32[0.011792993, 0.009161911, 0.009117997]
```
If a charge would hit an uncontacted surface of the semiconductor, the charge drift vector is projected parallel to the surface and the charge drift can be slowed down. To my knowledge, this warning appears when a charge drift would be orthogonal to the surface. Then, the projection of the charge drift step parallel to the surface is not unique. In these events, charge carriers might not reach any of the contacts but be stuck on the uncontacted surface, leading to a decrease in the waveform amplitude.


I know that these warnings might be printed often, especially when simulating a huge amount of events. In order to suppress them, have a look at [Suppressor.jl](https://github.com/JuliaIO/Suppressor.jl). You might run this as `@suppress_err simulate_waveforms`

**Reply by zql2021** [reply]
Thank you very much for your detailed answer, which resolved my confusion.
I've  another small question.
If I  set the grid precision too fine，**calculating electric field, electric potential, and weighting potential (or run simulate!() directly )will take a long time** .
So，I'm curious **if SolidStateDetectors has or can provide the function to save the calculated `sim`.**  I can load` sim` directly next time.

**Reply by fhagemann** [reply]
Yes, of course (and I would highly recommend to do so). You can find some information on how to save simulations [here](https://juliaphysics.github.io/SolidStateDetectors.jl/stable/man/IO/#IO).

I would recommend to use [LegendHDF5IO.jl](https://github.com/legend-exp/LegendHDF5IO.jl).

Run this once:
```
using Pkg
Pkg.add(url="https://github.com/legend-exp/LegendDataTypes.jl.git")
Pkg.add(url="https://github.com/legend-exp/LegendHDF5IO.jl.git")
Pkg.add("HDF5")
```

Then, you can run your field calculation once and save it using
```
using LegendHDF5IO
using SolidStateDetectors
sim = Simulation(...)
simulate!(...)
ssd_write("my_simulation.lh5", sim)
```
and read it in another Julia session using
```
using LegendHDF5IO
using SolidStateDetectors
sim = ssd_read("my_simulation.lh5", Simulation)
```

**Reply by zql2021** [reply]
That's so great,thank you for your guidance!
